### PR TITLE
Auth/middleware/abdi

### PR DESCRIPTION
**Pull request by abdimohamud**
# PR Description
​
Implemented custom middleware that supports RBAC (Role-Based Access Control) basics for (guests, supervisors, case-managers). Middleware checks role of user on endpoints implemented on and if permission is granted, lets them continue. Returns a error message if user doesn't have role/permission for endpoint. Each Role extends the rights and permission of the user below it, so a user with a higher role also has the same permissions for CRUD operations of a user with less permissions. This lets case managers view/add own notes and guests only view notes, supervisors & case managers view all families members, etc.  
​
​
## Type of Change
​
Types of Change:
- [x] New feature
- [x] Refactoring / Bug fix
- [] Documentation update 
​
​
## Testing
​
- [x] Yes
- [] No
- [] Not needed
​
​
## Completeness
Is this feature complete
- [x] Feature Complete
- [] Feature Incomplete, user story/trello card linked
​
# PR Requirements
- [x] Clean and concise commenting
- [x] Clean and comprehensive documentation included
- [x] Conforms to engineering standards and checks
- [x] There are no merge conflicts
- [x] There are no warnings
- [x] Corresponding documentation elsewhere has been updated to reflect these changes
​
​
## Requester Questions
​
```Add any questions, comments, or concerns you have as the person initiating the pull request for your reviewers```
- Might need refactoring towards what a user role should automatically enable them to view/update/create/delete, made new feature based off API contracts in design architecture. 
- Using drier code could be an improvement, chose to be safe in case certain tables required additional access for a certain role  
- Updating a role permission is easy, so if there is refactoring needed, should be a quick fix
​
​
## Other
[Rubric](https://www.notion.so/1fc04e4fedeb429ba873b7c68d281707?v=74054da7991341c0bf970f39410c43da)  
[Template Guidelines](https://www.notion.so/Pull-Request-Template-f9264f79e1b649b9845961b5aba3eaff)
[AccessRole Docs](https://onury.io/accesscontrol/?api=ac#AccessControl)
​

### CODE REVIEW

File: api/families/familiesMiddleware.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    return res.status(401).json({
+     error: "You don't have enough permission to perform this action"
+    });
+   }
+   next()
+  } catch (error) {
+   next(error)
+  }
+ }
+}
```
Comment (thereactgirl) [inline api/families/familiesMiddleware.js:17]: I really love how easy to understand this middleware is! 
especially this:
```const permission = roles.can(req.user.role)[action](resource);```
defining permission based on what the role is, 
what action is being made,
 and what resource is being accessed... **_chefs kiss_** 
 
I don't even have to look at the `roles` file to know what this is doing.

File: api/families/familiesRoles.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ .extend("guest")
+ .extend("case-manager")
+ .extend("supervisor")
+ .readAny("families")
+ .updateAny("families")
+ .deleteAny("families")
+ .createAny("families")
+ 
+return ac;
+})();
```
Comment (thereactgirl) [inline api/families/familiesRoles.js]: The `AccessControl` library is a great find! Kudos!

File: api/guestLogs/logsMiddleware.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    return res.status(401).json({
+     error: "You don't have enough permission to perform this action"
+    });
+   }
+   next()
+  } catch (error) {
+   next(error)
+  }
+ }
+}
```
Comment (thereactgirl) [inline api/guestLogs/logsMiddleware.js:17]: Does this mean a guest can't check-in? A guest should be able to create a log and a supervisor should be able to update a log with a `on_sight` value.

File: api/guestLogs/logsRoles.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
+// Supervisors should be able to view all guests currently checked into the shelter.
+const AccessControl = require("accesscontrol");
+const ac = new AccessControl();
+ 
+exports.roles = (function() {
+ac.grant("guest")
+  .createOwn("logs")
+ .readOwn("logs")
+ .updateOwn("logs")
```
Comment (thereactgirl) [inline api/guestLogs/logsRoles.js]: Ahh I see I don't think a guest should be able to update the log once they have checked in. @abdimohamud perhaps this should be changed they could most certainly decide not to check in, after all, Maybe they found somewhere to stay and they can delete their check in. @isaiah-j do you agree?

File: api/guestLogs/logsRoles.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+ ac.grant("case-manager")
+ .extend("guest")
+ .readAny("logs")
+ .updateAny("logs")
+ 
+ac.grant("supervisor")
+.extend("case-manager")
+ .extend("guest")
```
Comment (thereactgirl) [inline api/guestLogs/logsRoles.js]: case manager looks good but lets give the supervisor ability to update and delete logs

**Review by thereactgirl — APPROVED** [review]

File: api/guestLogs/logsRoles.js (hunk 4 above)
Comment (abdimohamud) [reply api/guestLogs/logsRoles.js]: good point, will need to take another look at the overall flow of the guest user story and revise some permissions so it makes sense for the what they can update and change on the tables.

File: api/guestLogs/logsRoles.js (hunk 5 above)
Comment (abdimohamud) [reply api/guestLogs/logsRoles.js]: will do, going to need this branch pushed soon in time for when new okta rolls in so we can test the capabilities and adjust to what makes sense

**Review by Nicci498 — APPROVED** [review]

**Pull request merged**